File: Count Collisions on a Road/Count Collisions on a Road.py
# Solution for Count Collisions on a Road in PY

class Solution:
    def countCollisions(self, directions: str) -> int:
        totalCars = len(directions)

        # Step 1: skip leading L cars
        leftIndex = 0
        while leftIndex < totalCars and directions[leftIndex] == 'L':
            leftIndex += 1
        
        # Step 2: skip trailing R cars
        rightIndex = totalCars - 1
        while rightIndex >= 0 and directions[rightIndex] == 'R':
            rightIndex -= 1
        
        # Step 3: count collisions in the middle part
        collisionCount = 0
        while leftIndex <= rightIndex:
            if directions[leftIndex] != 'S':
                collisionCount += 1
            leftIndex += 1
        
        return collisionCount

File: Count Collisions on a Road/Count Collisions on a Road.java
// Solution for Count Collisions on a Road in JAVA

class Solution {
    public int countCollisions(String directions) {
        int totalCars = directions.length();

        // Step 1: Skip all leading 'L' cars
        int leftIndex = 0;
        while (leftIndex < totalCars && directions.charAt(leftIndex) == 'L') {
            leftIndex++;
        }

        // Step 2: Skip all trailing 'R' cars
        int rightIndex = totalCars - 1;
        while (rightIndex >= 0 && directions.charAt(rightIndex) == 'R') {
            rightIndex--;
        }

        // Step 3: Count all non-'S' cars in the middle section
        int collisionCount = 0;
        while (leftIndex <= rightIndex) {
            if (directions.charAt(leftIndex) != 'S') {
                collisionCount++;
            }
            leftIndex++;
        }

        return collisionCount;
    }
}
